compiler/damlc/tests/src/DA/Test/DamlcIntegration.hs

```diff
@@ -331,77 +362,129 @@ getIntegrationTests registerTODO scenarioService (packageDbPath, packageFlags) =
             shutdown
             $ \service ->
           testGroup ("Tests for Daml-LF " ++ renderPretty version) $
-            map (testCase version isScriptV2Opt evalOrder service outdir registerTODO) damlTests
+            map (damlFileTestTree version isScriptV2Opt evalOrder service outdir registerTODO) damlTests
 
     pure tree
 
 newtype TestCase = TestCase ((String -> IO ()) -> IO Result)
 
 instance IsTest TestCase where
-  run _ (TestCase r) _ = do
-    logger <- newIORef DList.empty
-    let log msg = modifyIORef logger (`DList.snoc` msg)
-    res <- r log
-    msgs <- readIORef logger
-    let desc
-          | null (resultDescription res) = unlines (DList.toList msgs)
-          | otherwise = unlines (DList.toList (msgs `DList.snoc` resultDescription res))
-    pure $ res { resultDescription = desc }
+  run _ (TestCase r) _ =
+    withLog \log -> do
+      res <- r log
+      log (resultDescription res)
+      pure \msgs -> res { resultDescription = msgs }
   testOptions = Tagged []
 
-testCase :: LF.Version -> IsScriptV2Opt -> EvaluationOrder -> IO IdeState -> FilePath -> (TODO -> IO ()) -> (String, FilePath, [Ann]) -> TestTree
-testCase version (IsScriptV2Opt isScriptV2Opt) evalOrderOpt getService outdir registerTODO (name, file, anns) = singleTest name . TestCase $ \log -> do
-  service <- getService
-  if any (`notElem` supportedOutputVersions) [v | UntilLF v <- anns] then
-    pure (testFailed "Unsupported Daml-LF version in UNTIL-LF annotation")
-  else if any (ignoreVersion version) anns
-    then pure $ Result
-      { resultOutcome = Success
-      , resultDescription = ""
-      , resultShortDescription = "IGNORE"
-      , resultTime = 0
-      , resultDetailsPrinter = noResultDetails
+-- | 'withLog' takes a function 'f :: (String -> IO ()) -> IO (String -> a)',
+-- that is, a function such that given a logger function ':: String -> IO ()'
+-- returns (in an IO context) a function ':: String -> a' that constructs
+-- the final result in terms of the final state of the log.
+withLog :: ((String -> IO ()) -> IO (String -> a)) -> IO a
```

Comment:
I find the type of `withLog` a bit hard to understand so I tried to come up with alternative ideas but the best I could find is not much better. Writing it here in case you find it clearer.

Instead of `withLog` you could have a function `newLogger` that returns a pair of a `String -> IO ()` and an `IO String`:

```
data Logger = Logger { write :: String -> IO (), read :: IO String }
newLogger :: IO Logger
```
The `write` component writes to an `IORef (DList String)`, and the `read` component unlines the content of the IORef. That way, control is no longer inverted, and the code of the clients is arguably slightly more readable. But then you don't enforce anymore that the logs are only read after the action has been performed, which your `withLog` function ensures. Also it is a more verbose, so really, not very satisfactory.

Reply:
After sleeping on it, I think this type is fine. I wouldn't introduce the type aliases. If you squint and ignore the `IO`s, it is pretty clear: "I give you a logging function, you give me back a continuation/callback". It's almost `ContT (ReaderT FinalLog IO) Logger` but it's not quite!

Reply:
😅 before seeing your reply I gave up on trying to make a good docstring and went back to the `runWriterT`-like signature with just one inversion of control

Reply:
Looks good to me too :)